pairing.py

```diff
@@ -5,64 +5,48 @@
 
 from random import choices
```

Comment:
nit: why do we have choices from random and `np.random.choice(...)`

Can we just use numpy's random class?

Reply:
`np.random.choice` only samples uniformly

Reply:
it has a weighted version: https://numpy.org/doc/stable/reference/random/generated/numpy.random.choice.html

the argument is `p`